ADR/0018-apps-continuous-perf-testing.md

```diff
@@ -0,0 +1,72 @@
+>
+> - Average page load latency.
+> - Backend service CPU and memory usage...
+> - PostgreSQL CPU and memory usage…
+> - Internal PostgreSQL metrics like number of commits during the test or average number of database sessions during the test…
+>
+> If any of these metrics does not align to historical results, we mark this test result as a failure.
+
+And even if some test fails, the application team should be able to review the data and change the test definition, so next time a new test result is being evaluated based on the historical results, this new result will be one of them.
+
+The algorithm that says if the current test passes or fails when compared to historical results can vary. It can be:
+
+- Checking minimum and maximum of the metric in historical data and making sure the new result is in this range.
+- Using a floating average of historical data and setting a safe range around it and making sure the new result is in that range.
+- Some machine learning thingy.
+
+Goal of this ADR is to propose a way for this kind of testing to be implemented in Stonesoup (feature [STONE-679](https://issues.redhat.com/browse/STONE-679)). Even if it would not deprecate full-blown performance and scale testing, having a release gate with some short and small scale performance test is desirable for many application teams.
+
+We are considering two different options here. They differ in what portion of logic mentioned above is executed where.
+
+### Glossary
+
+- SUT: Subject Under Test
+- Horreum: a results storage service: <https://github.com/Hyperfoil/Horreum>
+- CPT: Continuous Performance Testing
+
+## Decision
+
+Let's use this architecture with a single Horreum instance that would be managed by Red Hat and used by all customers.
+
+![Architecture diagram with Horreum](assets/0018-apps-continuous-perf-testing.svg "Architecture diagram with Horreum")
+
+1. Performance test runs in the Tekton pipeline and generates JSON with test parameters and results.
+2. We gather configured monitoring metrics and add them to the JSON.
+3. We upload the JSON with all the data to Horreum.
+4. Horreum performs result analysis, looking for changes in configured metrics.
+4. We get PASS/FAIL decision from Horreum back to the pipeline, so pipeline can rerurn proper result.
+
+## Consequences
+
+Pros:
+
+- We can provide some best practices to customers wanting to do some basic performance regression testing.
+- Horreum is end-to-end solution that already exists, is already used by multiple teamsn and has documentation for the whole process.
+- StoneSoup UI can display all the trends because historical data will be available in Horreum.
+- Red Hat team that develops Horreum is willing to help with this effort.
+- Integration scripts that would be running in the pipeline (to gather monitoring data, to upload results to Horreum, to get PASS/FAIL decision from Horreum...) mostly exists.
```

Comment:
What additional work would be needed as part of this? How much of the additional work here is already "available"? How much needs to be developed? How much is customized to Stonesoup?

Reply:
I will ask John O'Harra to answer this.

Reply:
There are existing tools/scripts to monitor performance workloads and process the data in Horreum.  The additional work required would be to integrate into Stonesoup;
 - providing the data required for the Stonesoup UI 
 - determine how users configure the change detection parameters for their tests
 - document / provide a step for integrating the change detection into their pipeline (e.g. for Jenkins we have a Horreum plugin that allows users to easily upload their data)

Reply:
Instead of indicating that scripts mostly exist, can we just highlight what _does_ exist here and then in the `cons`, highlight what _does not_ exist?

Reply:
I see. Right. Added what John wrote above in d244174.